### ISSUE TITLE
Errors not propagated properly

### ISSUE THREAD

**Issue by dfujim** [dfujim/bfit#111]
In the case of constrained parameters, the original parameter errors are calculated simply as `fn(error)`, without any proper error propagation. 

One possible solution: calculate errors using MC simulation as was done on the [mcerror](https://github.com/dfujim/bfit/tree/mcerror) branch. 

This is very slow, even with parallelization. Beyond implementation issues, of which there are many, about `1e6` random samples are needed to properly estimate the propagated error. This takes about 12 seconds for serial operation, or 3 seconds on 4 processes using `pandarallel`. Times two (pos and neg errors) for each run and each parameter constrained, this can quickly get out of hand. Ex: serial operation with one constrained parameter and 5 runs = 2 min wait time just for the errors to be calculated.

There is also an additional error where for some unknown reason, bfit cycles though the runs and parameters at least twice when updating the values. So the above example should take at least 4 min. May there be a loop somewhere with traces calling each other? Unsure. 

As an aside: the variable traces are called on each populate call resulting in a triggering of this calculation every time the constraint button is pressed. So this 4 min wait time is induced even without fitting. I don't think this is a tenable solution in this state.

**Comment by dfujim** [comment]
@rmlmcfadden ideas are appreciated

**Comment by rmlmcfadden** [comment]
Some thoughts:

- I'm surprised it's so slow! I would have also thought that fewer samples would have been sufficient for a "good enough" error estimate. It would be interesting to see a plot of `fn_df.std` vs. <i>N</i><sub>samples</sub>. 
- I wonder if some time could be shaved off by only saving the `fn` evaluations in `get_mcerror`? That is, generating the random input for `fn` individually rather than storing <i>all</i> of it and doing the evaluations after the fact. This avoids having to create the <i>N</i><sub>par</sub> x 10<sup>6</sup> `Dataframe`, which I'm sure has a bit of overhead. Not exactly a "Pythonic" solution, but it may help...
- I think it's fair to forgo the asymmetric error bars entirely, especially if it just doubles the runtime.
- Can [`JAX`](https://github.com/google/jax) help with this?

**Comment by dfujim** [comment]
* I started with N = 10<sup>5</sup>, but for any given input, running the MC simulation gave different answers. Thus unstable. I therefore bumped it up to N =10<sup>6</sup> and got better results.

* Yes, there are certainly some optimization to be had (notes [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/enhancingperf.html)). My understanding was that pandas `DataFrame`s are implemented in C and should thus be fast. So I was using the `apply` function as a way of getting around using a for loop, which would certainly be slow. Pandas should have no problem with arrays on the order of 10<sup>6</sup>. Note:
```python
In: %timeit pd.Series(np.arange(1e6))                                                                                           
Out: 935 µs ± 25 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
The bottleneck appears to be calculating the (python) function evaluations. While writing this, I was looking into the source of the bottleneck and found out that pandas.apply is slow. Simply using numpy arrays is much faster (see 1956ad40dadfe28af7c62f571e47bc2ab738d3f5), but there is still a worryingly large amount of variance in the output: 

```python
In: np.std([get_mcerror(lambda a,b : a+b, (1, 2), (0.1, 0.5), n=1e6) for i in range(1000)])
Out: 0.0003586249988422781

In: np.std([get_mcerror(lambda a,b : a+b, (1, 2), (0.1, 0.5), n=1e7) for i in range(100)])
Out: 0.00012225699621276
```

* No, I disagree simply because of some people who very much want to see minos errors. The run time for this calculation should not pose an issue, ideally. 

* I've started looking at JAX, which seems to be a great solution (see branch [derror](https://github.com/dfujim/bfit/tree/derror)), except in the case where numpy functions are used. It currently crashes, presumably because some input somewhere is passing a list instead of an array. Once I fix this, I think this will be a great solution.

**Comment by dfujim** [comment]
I was documenting the problem here and stumbled upon the solution. I have preserved my investigation for posterity. 

### Update on jax implementation

While this works outside of bfit (frustrating), the issue appears to be with `scipy.curve_fit`. I have no idea why this works outside of pressing the fit button. Even calling the functions and `curve_fit` from the "Run Commands" window produces something reasonable. 

When fitting, `curve_fit` seems to attempt to determine the gradient in the constrained parameter by sending input `1.` and then `1.00000001`. The `DeviceArray` object that jax uses doesn't seem to track enough decimal places to tell the difference.  

In [functions.py](https://github.com/dfujim/bfit/blob/043fd904a7becf70ce879402ffdb6434c3550852/bfit/fitting/functions.py), insert the indicated print statement: 
```python
  def new_fn(x, *pars):
      pars = np.asarray(pars)     
      new_inpt = [i[0](*pars[i[1]]) if type(i) is tuple else pars[i] for i in inpt]
      print(pars, new_inpt)   # line 77 insert this code
      return fn(x, *new_inpt)
```
to get 

```
[0.127 1.   ] [DeviceArray(1., dtype=float32), 0.127]
[0.12700001 1.        ] [DeviceArray(1., dtype=float32), 0.1270000149011612]
[0.127      1.00000001] [DeviceArray(1., dtype=float32), 0.127]
[1.13503224 1.        ] [DeviceArray(1., dtype=float32), 1.1350322415478584]
[0.37900806 1.        ] [DeviceArray(1., dtype=float32), 0.3790080603869646]
[0.19000202 1.        ] [DeviceArray(1., dtype=float32), 0.19000201509674114]
[0.1427505 1.       ] [DeviceArray(1., dtype=float32), 0.1427505037741853]
[0.13093763 1.        ] [DeviceArray(1., dtype=float32), 0.13093762594354633]
[0.12798441 1.        ] [DeviceArray(1., dtype=float32), 0.12798440648588658]
[0.12798442 1.        ] [DeviceArray(1., dtype=float32), 0.12798442138704777]
[0.12798441 1.00000001] [DeviceArray(1., dtype=float32), 0.12798440648588658]
# etc
```

Note that the order of the two arrays are inverted (purposefully). Experiment: 

```python
In:  import jax.numpy as jnp
In:  jnp.exp(1.) == jnp.exp(1.00000001)                                                              
Out: DeviceArray(True, dtype=bool, weak_type=True)
```

Oddly enough, `jax.numpy` doesn't seem to have a very good implementation of Euler's number: 

```python
import jax.numpy as jnp
import numpy as np

print(' np', float(np.exp(1)))
print('jnp', float(jnp.exp(1)))
```

has as output

```
 np 2.718281828459045
jnp 2.7182817459106445
```

It's likely not a coincidence that `float32` has an accuracy of about 7 decimal places, which is where the numbers start to differ (`jax.numpy` uses `float32` whereas `numpy` uses `float64`). 

### Solution 

This was exactly the solution. Add the following line to [popup_fit_constraints.py](https://github.com/dfujim/bfit/blob/master/bfit/gui/popup_fit_constraints.py)

```python 
jax.config.update("jax_enable_x64", True)
```
to enable the use of 64 bit floats, thus tracking up to 16 decimal places of accuracy. 

Updated in 7a83a37a672864280fee2a6be98c66f0e2f67a1f

**Comment by dfujim** [comment]
Merged branches and pushed update. Went with the jax solution.
